File: C++/Graph/Bellman-ford/Bellman-ford.cpp
#include<bits/stdc++.h>

using namespace std;

typedef long long ll;

const int MAX_N = 5e2 + 2;
const int MAX_M = 6e3 + 3;
const ll INF =  0x3f3f3f3f3f3f3f3f;

int n, m;   // n = number of node, m = number of edge

struct Edge{
    int u,v;    // node1, node2
    ll w;       // weight
};

Edge edge[MAX_M];
ll dist[MAX_N];


bool bellmanFord(int start)
{
    bool isCycle = false;
    memset(dist, 0x3f, sizeof(dist));
    dist[start] = 0;
    
    // repeat multiple rounds to determine if a negative cycle exists
    for(int round = 0; round < m; ++round){
        bool updated = false;
        
        for(auto& e: edge)
        {
            if(dist[e.u] != INF && dist[e.u] + e.w < dist[e.v]){
                dist[e.v] = dist[e.u] + e.w;
                updated = true;
                
                // m - 1 repeat means that the shortest path continues to be updated
                // there is a negative cycle
                if(round == m-1)
                    isCycle = true;
            }
        }
        if (!updated) break;
    }
    
    return isCycle;
}

int main()
{
    ios_base::sync_with_stdio(0);
    cin.tie(0);

    int u, v;
    ll w;
    
    cin >> n >> m;
    
    for(int i=0; i<m; ++i){
        cin >> u >> v >> w;
        edge[i] = {u,v,w};
    }
    
    int start = 1;
    bool negativeCycle = bellmanFord(start);
    
    if(negativeCycle) cout << -1 << "\n";
    
    else{
        for(int i=1; i<=n; ++i) if(i!=start){
            cout << (dist[i] == INF ? -1 : dist[i]) << '\n';
        }
    }
    
    return 0;
}

File: Python/Graph/Bellman-ford/Bellman-ford.py
def bellman_ford(start):
    global dist, edge

    is_cycle = False
    dist = [INF for _ in range(n + 1)]
    dist[start] = 0

    # repeat multiple rounds to determine if a negative cycle exists
    for round in range(m):
        updated = False

        for e in edge:
            src, dst, weight = e
            if dist[src] is not INF and dist[src] + weight < dist[dst]:
                dist[dst] = dist[src] + weight
                updated = True

                # m - 1 repeat means that the shortest path continues to be updated
                # there is a negative cycle
                if round == m - 1:
                    is_cycle = True

        if not updated:
            break

    return is_cycle


if __name__ == '__main__':
    INF = 0x3f3f3f3f3f3f3f3f

    tc = int(input())

    for i in range(tc):
        edge = []
        dist = []

        n, m = map(int, input().split())

        for _ in range(m):
            u, v, w = map(int, input().split())
            edge.append((u, v, w))

        negative_cycle = bellman_ford(1)

        if i == 0:
            print()

        print("test case: ", i + 1)
        if negative_cycle:
            print("negative cycle exists")

        else:
            for j in range(2, n + 1):
                print("dist[{}]: {}".format(j, dist[j]) if (dist[i] is not INF) else -1)
